ampd/src/handlers/xrpl_verify_msg.rs

```diff
@@ -0,0 +1,195 @@
+use std::collections::{HashMap, HashSet};
+use std::convert::TryInto;
+use std::str::FromStr;
+
+use async_trait::async_trait;
+use axelar_wasm_std::msg_id::HexTxHash;
+use axelar_wasm_std::voting::{PollId, Vote};
+use cosmrs::cosmwasm::MsgExecuteContract;
+use cosmrs::tx::Msg;
+use cosmrs::Any;
+use error_stack::ResultExt;
+use events::Error::EventTypeMismatch;
+use events_derive::try_from;
+use futures::future::join_all;
+use router_api::ChainName;
+use serde::Deserialize;
+use tokio::sync::watch::Receiver;
+use tracing::{info, info_span};
+use valuable::Valuable;
+use voting_verifier::msg::ExecuteMsg;
+use xrpl_http_client::Transaction;
+use xrpl_types::msg::XRPLMessage;
+use xrpl_types::types::{xrpl_account_id_string, XRPLAccountId};
+
+use crate::event_processor::EventHandler;
+use crate::handlers::errors::Error;
+use crate::types::TMAddress;
+use crate::xrpl::finalizer::{self, Finalization};
+use crate::xrpl::json_rpc::XRPLClient;
+use crate::xrpl::verifier::verify_message;
+
+type Result<T> = error_stack::Result<T, Error>;
+
+#[derive(Deserialize, Debug)]
+#[try_from("wasm-messages_poll_started")]
+struct PollStartedEvent {
+    poll_id: PollId,
+    source_chain: ChainName,
+    #[serde(with = "xrpl_account_id_string")]
+    source_gateway_address: XRPLAccountId,
+    expires_at: u64,
+    messages: Vec<XRPLMessage>,
+    participants: Vec<TMAddress>,
+}
+
+pub struct Handler<C>
+where
+    C: XRPLClient + Send + Sync,
+{
+    verifier: TMAddress,
+    voting_verifier_contract: TMAddress,
+    finalizer_type: Finalization,
+    rpc_client: C,
+    latest_block_height: Receiver<u64>,
+}
+
+impl<C> Handler<C>
+where
+    C: XRPLClient + Send + Sync,
+{
+    pub fn new(
+        verifier: TMAddress,
+        voting_verifier_contract: TMAddress,
+        finalizer_type: Finalization,
+        rpc_client: C,
+        latest_block_height: Receiver<u64>,
+    ) -> Self {
+        Self {
+            verifier,
+            voting_verifier_contract,
+            finalizer_type,
+            rpc_client,
+            latest_block_height,
+        }
+    }
+
+    async fn validated_txs<T>(&self, tx_ids: T) -> Result<HashMap<HexTxHash, Transaction>>
+    where
+        T: IntoIterator<Item = HexTxHash>,
+    {
+        let latest_validated_ledger_index = finalizer::pick(&self.finalizer_type, &self.rpc_client)
+            .latest_validated_ledger_index()
+            .await
+            .change_context(Error::Finalizer)?;
+
+        Ok(join_all(
+            tx_ids
+                .into_iter()
+                .map(|tx_id| self.rpc_client.tx(tx_id.tx_hash)),
+        )
+        .await
+        .into_iter()
+        .filter_map(std::result::Result::unwrap_or_default)
+        .filter_map(|tx_res| {
+            let tx_common = tx_res.tx.common();
+            let (ledger_index, tx_hash) = (tx_common.ledger_index?, tx_common.hash.clone()?);
+
+            if ledger_index.ge(&latest_validated_ledger_index) {
+                return None;
+            }
```

Comment:
you don't need this check. You can just check for the `validated` field in the `tx` response. Will save you an rpc call per message.

Reply:
yep, removed: https://github.com/axelarnetwork/axelar-amplifier/pull/764/commits/fe41627e7bad6c7c35250db5504945452fd86e83